misc/config_tools/launch_config/launch_cfg_gen.py

```diff
@@ -1,269 +1,366 @@
+
+    # Mediated PCI devices, including virtio
+    for usb_xhci in vm_launch_etree.xpath(".//usb_xhci[text() != '']/text()"):
+        script.add_virtual_device("xhci", options=usb_xhci)
+
+    for virtio_input in vm_launch_etree.xpath(".//virtio_devices/input[text() != '']/text()"):
+        script.add_virtual_device("virtio-input", options=virtio_input)
+
+    for virtio_console in vm_launch_etree.xpath(".//virtio_devices/console[text() != '']/text()"):
+        script.add_virtual_device("virtio-console", options=virtio_console)
+
+    for virtio_network in vm_launch_etree.xpath(".//virtio_devices/network[text() != '']/text()"):
+        params = virtio_network.split(",", maxsplit=1)
+        tap_conf = f"tap={params[0]}"
+        params = [tap_conf] + params[1:]
+        script.add_init_command(f"mac=$(cat /sys/class/net/e*/address)")
+        params.append(f"mac_seed=${{mac:0:17}}-{vm_name}")
+        script.add_virtual_device("virtio-net", options=",".join(params))
+
+    for virtio_block in vm_launch_etree.xpath(".//virtio_devices/block[text() != '']/text()"):
+        params = virtio_block.split(":", maxsplit=1)
+        if len(params) == 1:
+            script.add_virtual_device("virtio-blk", options=virtio_block)
+        else:
+            block_device = params[0]
+            rootfs_img = params[1]
+            var = f"dir_{os.path.basename(block_device)}"
+            script.add_init_command(f"{var}=`mount_partition {block_device}`")
+            script.add_virtual_device("virtio-blk", options=os.path.join(f"${{{var}}}", rootfs_img))
+            script.add_deinit_command(f"unmount_partition ${{{var}}}")
+
+    # Passthrough PCI devices
+    bdf_regex = re.compile("([0-9a-f]{2}):([0-1][0-9a-f]).([0-7])")
+    for passthru_device in vm_launch_etree.xpath(".//passthrough_devices/*/text()"):
+        m = bdf_regex.match(passthru_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16))
+
+    for sriov_gpu_device in vm_launch_etree.xpath(".//sriov/gpu/text()"):
+        m = bdf_regex.match(sriov_gpu_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16), options="igd-vf")
+
+    ###
+    # Miscellaneous
+    ###
+    script.add_dynamic_dm_parameter("add_rtvm_options")
+    script.add_dynamic_dm_parameter("add_logger_settings", "console=4 kmsg=3 disk=5")
+
+    ###
+    # Lastly, conclude the device model parameters with the VM name
+    ###
+    script.add_plain_dm_parameter(f"{vm_name}")
+
+    return script
+
+def main(board_xml, scenario_xml, launch_xml, user_vmid, out_dir):
+    board_etree = etree.parse(board_xml)
+    scenario_etree = etree.parse(scenario_xml)
+    launch_etree = etree.parse(launch_xml)
+
+    sos_vm_id = eval_xpath(scenario_etree, "//vm[vm_type='SERVICE_VM']/@id")
```

Comment:
Can we avoid using `sos` here, e.g. change that variable name to `service_vm_id`?

Reply:
Updated in v3.